doc/mmlicense.1
.\"@mindmaze_header@
.TH MMLICENSE 1 2013 "MINDMAZE" "mmlib library manual"
.SH NAME
mmlicense - program to manage the hardware authorization token
.SH SYNOPSIS
.SY mmlicense 
.OP \-\-cert file
.OP \-\-privkey file
.OP \-\-signature-dir dir
.OP \-\-hw\-file file
.RB { allow | check }
.YS
.SY mmlicense 
.RB { \-\-help | \-h }
.YS
.SH DESCRIPTION
.LP
.B mmlicense
is the software that manages the hardware authorization tokens. Those tokens
allow to run Mindmaze software on a hardware and are generated from a X509
certificate. More precisely each token corresponds to a signature of the
hardware description by a certificate recognized by the Mindmaze Certificate
Authority (CA).
.LP
Unless the \fB\-h\fP, or \fB\-\-help\fP option is given, one of the commands
below must be present:
.TP 4
.B allow
Generate an authorization token for the current hardware. The X509
certificate and its private key used for signature are expected to be
located respectively at ~/.config/mindmaze/default.crt and
~/.config/mindmaze/default.key. The generated token will be placed in the
folder ~/.config/mindmaze. Use
.BR pki-gen-client-cert (1)
if you do not have already a Mindmaze certificate.
.
.TP
.B check
The authorization token verification as if it were run at execution of
a mindmaze software. If no token for the host hardware is found either in
/etc/mindmaze or ~/.config/mindmaze folder or if the signature bundled in
the token does not match the hardware info or if the certificate is invalid
in regard of the mindmaze CA, then the check will fail.
.LP
.SH OPTIONS
.TP 4
.BI \-\-cert= file
Use the specified \fIfile\fP for the X509 certificate for generating the
authorization token.
.
.TP
.BI \-\-privkey= file
Use the specified \fIfile\fP for the private key for generating the
authorization token. It must be the private key corresponding to the public
key bundled in the specified X509 certificate (be it implicitely or
explicitely set with the \fB\-\-cert\fP option).
.
.TP
.BI \-\-signature-dir= dir
Place the generated token in the \fIdir\fP folder instead of
~/.config/mindmaze.
.
.TP 
.BI \-\-hw\-file= file
Use the specified hardware info to generate a token instead of the hardware
of the host system.
.
.TP
.B \-\-help|\-h
Display the command-line help
.SH ENVIRONMENT
.TP 4
.B XDG_CONFIG_HOME
If set, all the path relative to ~/.config are replaced using
$XDG_CONFIG_HOME.
.TP
.B MM_LIC_ALTDIR
If set, the license check will search for a valid token in the folder
specified by $MM_LIC_ALTDIR. This search in done addition to the
search perform in the usual folder. This environment variable is intended
for automated build system. It is worth noting that although this variable
is mentioned in
.B mmlicense
manpage, it is of course understood by the mmlib library when the token
check is performed at runtime.
.SH "SEE ALSO"
.BR pki-gen-client-cert (1)



debian/changelog
mmlib (0.0.3) testing; urgency=low

  * New module version:
    - Fix memory leak in read signature

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Fri, 06 Sep 2013 12:46:39 +0200

mmlib (0.0.2) testing; urgency=low

  * New module version:
    - robustify build
  * Declare breakage of old libmmlib0 by libmmlib-dev
  * Add missing build-deps of gnutls-bin
  * Re-enable unit tests during package building

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Mon, 02 Sep 2013 08:39:18 +0200

mmlib (0.0.1) testing; urgency=low

  * New module version:
    - Add license check
  * Add libgnutls-dev and ca-certificates-mindmaze as build-depends
  * Install predefined tokens with libmmlib0
  * Install mmlicense in mmlib-bin package
  * Install development manpages in libmmlib-dev
  * Add dependency on pciutils
  * Disable unit test during package building

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Fri, 30 Aug 2013 11:32:55 +0200

mmlib (0.0.0) testing; urgency=low

  * Added skeleton save/load
  * New error codes defined

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Thu, 22 Aug 2013 08:26:35 +0200

mmlib (0.0.0~~experimental5) prealpha; urgency=low

  * Added skeleton function

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Tue, 11 Jun 2013 15:28:07 +0200

mmlib (0.0.0~~experimental3) prealpha; urgency=low

  * Added angle axis - quaternion conversion

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Fri, 15 Mar 2013 18:40:52 +0100

mmlib (0.0.0~~experimental2) prealpha; urgency=low

  * new error codes (for database connection)

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Fri, 01 Mar 2013 10:38:50 +0100

mmlib (0.0.0~~experimental) prealpha; urgency=low

  * Initial debian package.

 -- Nicolas Bourdaud <nicolas.bourdaud@mindmaze.ch>  Thu, 24 Jan 2013 10:22:18 +0100


src/mmlicense.c
/*
   @mindmaze_header@
*/
#if HAVE_CONFIG_H
# include <config.h>
#endif

#define _XOPEN_SOURCE

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <errno.h>
#include <getopt.h>
#include <gnutls/gnutls.h>
#include <gnutls/abstract.h>
#include <gnutls/crypto.h>
#include <gnutls/x509.h>

#include "license.h"


char homecfg[128];

static const char* hwfile;
static char privkey[256];
static char sigdir[256];
static char certfile[256];


static
void init_globals(void)
{
	size_t len = sizeof(homecfg);
	const char* envvar = getenv("XDG_CONFIG_HOME");

	if (envvar)
		strncpy(homecfg, envvar, len);
	else
		snprintf(homecfg, len, "%s/.config", getenv("HOME"));
	
	sprintf(privkey, "%s/mindmaze/default.key", homecfg);
	sprintf(certfile, "%s/mindmaze/default.crt", homecfg);
	sprintf(sigdir, "%s/mindmaze", homecfg);
}


enum {
	CERT,
	PRIVKEY,
	SIGDIR,
	HWFILE,
};

static const struct option opt_str[] = {
	{"cert", required_argument, NULL, CERT},
	{"privkey", required_argument, NULL, PRIVKEY},
	{"signature-dir", required_argument, NULL, SIGDIR},
	{"hw-file", required_argument, NULL, HWFILE},
	{"help", 0, NULL, 'h'},
	{NULL}
};

static
void print_usage(const char* execname, int success)
{
	fprintf(stderr,	"Usage: \n"
	        "   %s [options] <allow|check> \n"
	        "options:\n"
	        "\t--cert=FILE         : path of the X509 certificate"
		" (default: %s/mindmaze/default.crt)\n"
	        "\t--privkey=FILE      : path of the private key"
		" (default: %s/mindmaze/default.key)\n"
		"\t--signature-dir=DIR : folder where the signature will "
	        "be put (default: %s/mindmaze)\n"
		"\t--hw-file=FILE     : the hw dump used for signature "
		"(default: output of 'lspci -n')\n"
		"\t--help | -h         : display this help\n",
		execname, homecfg, homecfg, homecfg);
	exit(success ? EXIT_SUCCESS : EXIT_FAILURE);
}


static
const char* parse_options(int argc, char *argv[])
{
	int c;
	int option_index = 0;

	while (1) {
		c = getopt_long(argc, argv, "h", opt_str, &option_index);
		if (c == -1)
			break;

		switch (c) {
		case CERT:	strcpy(certfile, optarg); break;
		case PRIVKEY:	strcpy(privkey, optarg); break;
		case HWFILE:	hwfile = optarg; break;
		case SIGDIR:	strcpy(sigdir, optarg); break;
		case 'h':
		case '?':
			print_usage(argv[0], (c == 'h') ? 1 : 0);
		}
	}

	if (optind+1 != argc) {
		fprintf(stderr, "Too %s arguments.\n",
		                    (optind+1 > argc) ? "few" : "many");
		print_usage(argv[0], 0);
	}

	return argv[optind];
}


int main(int argc, char* argv[])
{
	int r = 0;
	const char* cmd;

	init_globals();
	cmd = parse_options(argc, argv);

	gnutls_global_init();
	if (!strcmp("allow", cmd)) {
		gen_signature(hwfile, sigdir, certfile, privkey);
	} else if (!strcmp("check", cmd)) {
		r = check_signature(hwfile);
		fprintf(stderr, "signature verification %s\n",
		                r ? "failed" : "succeeded");
	} else {
		fprintf(stderr, "Unknown command: %s\n", cmd);
		r = -1;
	}
	gnutls_global_deinit();

	return (r == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
}

